### ISSUE TITLE
Psalm does not understand implicit null return values

### ISSUE THREAD

**Issue by BenMorel** [vimeo/psalm#6826]
Take the following example:

```php
function test(string ...$_values): void {}

test(
    ...array_map(
        function (string $x) {
            if (rand(0,1)) {
                return $x;
            }

            return null;
        },
        ['a', 'b', 'c'],
    )
);
```

https://psalm.dev/r/55219ce9a3

Psalm rightfully reports:

> ERROR: PossiblyNullArgument - 6:8 - Argument 1 of test cannot be null, possibly null value provided

If, however, you remove the explicit `return null`, Psalm does not complain anymore:

```php

test(
    ...array_map(
        function (string $x) {
            if (rand(0,1)) {
                return $x;
            }
        },
        ['a', 'b', 'c'],
    )
);
```

https://psalm.dev/r/ef09fc96a7

> No issues!

Even though the result is the same, as not returning anything implicitly returns `null`.

**Comment by weirdan** [comment]
Simplified: https://psalm.dev/r/a157fb2360

**Comment by orklah** [comment]
This is a duplicate of https://github.com/vimeo/psalm/issues/5179

and the reason behind my failed PR here: https://github.com/vimeo/psalm/pull/6644 and also two consecutive failed branches that didn't even made it to the PR stage because they were too broken!

I'll get it working eventually!

**Comment by BenMorel** [comment]
Thank you!
